<a href="https://colab.research.google.com/github/juyoung524/22-1-Spring-/blob/main/20220509.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Tokenization

In [1]:
import nltk

In [ ]:
# copy a file from github
import os
url = "https://raw.githubusercontent.com/hsnam95/class2022Spring/main/crime_punishment.txt"
os.system("curl " + url + " > crime_punishment.txt")

# read a text file in the server
file = open("crime_punishment.txt")
text = file.read().replace("\n", " ")
file.close()

In [ ]:
# or copy/pase text here
text = 'Here’s to the crazy ones, the misfits, the rebels, the troublemakers, the round pegs in the square holes. \
The ones who see things differently — they’re not fond of rules. \
You can quote them, disagree with them, glorify or vilify them, \
but the only thing you can’t do is ignore them because they change things. \
They push the human race forward, and while some may see them as the crazy ones, we see genius, \
because the ones who are crazy enough to think that they can change the world, are the ones who do.'

In [ ]:
# write out a text file
file = open("tmp.txt", "w")
file.write(text)
file.close()

In [ ]:
text.split()

In [ ]:
' '.join(text.split())

In [ ]:
from nltk.tokenize import word_tokenize
nltk.download('punkt')
words = word_tokenize(text)

In [ ]:
from nltk.tokenize import RegexpTokenizer
retokenize = RegexpTokenizer("[\w]+")
words = retokenize.tokenize(text)

### Normalization  
**Stemming** 어간 추출 대충의 패턴 규칙으로 어미를 잘라내는 것 (사전에 없는 어간 나올 수 있음)

**Lemmatization** 표제어(기본 사전형) 추출.


In [ ]:
from nltk.stem import PorterStemmer
stemmer = PorterStemmer()
[stemmer.stem(w) for w in words]

In [ ]:
from nltk.stem import LancasterStemmer
stemmer = LancasterStemmer()
[stemmer.stem(w) for w in words]

In [ ]:
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()
[lemmatizer.lemmatize(w) for w in words]

### Stopword

In [ ]:
from nltk.corpus import stopwords  
nltk.download('stopwords')
print(words)
words = [w for w in words if not w in stopwords.words('english')]
print(words)

### Collocation, Concordance

In [ ]:
nltk.download('gutenberg')
text = nltk.corpus.gutenberg.raw('austen-emma.txt')
words = retokenize.tokenize(text)

In [ ]:
nltk.Text(words).collocations()  # default: (num=20, window_size=2)

In [ ]:
nltk.Text(words).concordance('Emma', 79, 10)

In [ ]:
nltk.Text(words).dispersion_plot(["Emma", "Knightley", "Frank", "Jane", "Harriet", "Robert"])

In [ ]:
# Distributional similarity: 
# find other words which appear in the same contexts as the specified word; 
# list most similar words first.
nltk.Text(words).similar("Emma")

In [ ]:
# Find contexts where the specified words appear; list most frequent common contexts first.
nltk.Text(words).common_contexts(["Emma", "she"])

### Frequency distribution, Frequency plot

In [ ]:
fd = nltk.FreqDist(words).most_common(20)
fd

In [ ]:
nltk.Text(words).plot(20)

### Dictionary

In [ ]:
nltk.download('words')
nltk.corpus.words.words('en')[-20:-1]


[nltk_data] Downloading package words to /root/nltk_data...
[nltk_data]   Unzipping corpora/words.zip.


['zymosis',
 'zymosterol',
 'zymosthenic',
 'zymotechnic',
 'zymotechnical',
 'zymotechnics',
 'zymotechny',
 'zymotic',
 'zymotically',
 'zymotize',
 'zymotoxic',
 'zymurgy',
 'Zyrenian',
 'Zyrian',
 'Zyryan',
 'zythem',
 'Zythia',
 'zythum',
 'Zyzomys']

In [ ]:
len(nltk.corpus.words.words('en'))

### Extract information (pos tag, named entity)


#### **POS tag list**:

CC	coordinating conjunction \
CD	cardinal digit \
DT	determiner \
EX	existential there (like: "there is" ... think of it like "there exists") \
FW	foreign word \
IN	preposition/subordinating conjunction \
JJ	adjective	'big' \
JJR	adjective, comparative	'bigger' \
JJS	adjective, superlative	'biggest' \
LS	list marker	1) \
MD	modal	could, will \
NN	noun, singular 'desk' \
NNS	noun plural	'desks' \
NNP	proper noun, singular	'Harrison' \
NNPS	proper noun, plural	'Americans' \
PDT	predeterminer	'all the kids' \
POS	possessive ending	parent's \
PRP	personal pronoun	I, he, she \
PRP\$	possessive pronoun	my, his, hers \
RB	adverb	very, silently, \
RBR	adverb, comparative	better \
RBS	adverb, superlative	best \
RP	particle	give up \
TO	to	go 'to' the store. \
UH	interjection	errrrrrrrm \
VB	verb, base form	take \
VBD	verb, past tense	took \
VBG	verb, gerund/present participle	taking \
VBN	verb, past participle	taken \
VBP	verb, sing. present, non-3d	take \
VBZ	verb, 3rd person sing. present	takes \
WDT	wh-determiner	which \
WP	wh-pronoun	who, what \
WP\$	possessive wh-pronoun	whose \
WRB	wh-abverb	where, when \

In [ ]:
sent = "I am Jhon from America and would like to go to Starbuck"
words = nltk.word_tokenize(sent)

In [ ]:
nltk.download('averaged_perceptron_tagger')
pos = nltk.pos_tag(words)

In [ ]:
nltk.download('maxent_ne_chunker')
NE = nltk.ne_chunk(pos)
# common Entity types: ORGANIZATION, PERSON, LOCATION, DATE, TIME, MONEY, and GPE (geo-political entity)

### Wordcloud

In [ ]:
from wordcloud import WordCloud, STOPWORDS
import matplotlib.pyplot as plt

text = nltk.corpus.gutenberg.raw('bible-kjv.txt')

wc = WordCloud().generate(text) 
plt.imshow(wc)

In [ ]:
stopwords = set(STOPWORDS) 
stopwords.add('unto')
wc = WordCloud(stopwords = stopwords).generate(text) 
plt.imshow(wc)

### Regular expression

In [2]:
import re

In [3]:
re.search('a', 'abcdefa')

<re.Match object; span=(0, 1), match='a'>

제일 먼저 나온 것을 찾아주는 것임

In [4]:
re.findall('a', 'abcdefa')

['a', 'a']

다 찾아서 list로 만들어주는 것 

(처음부터 잠깐 다시 듣기)

Regular expression 잘 쓰면 어떻게 유용하게 쓰이는지 알 수 있을 것임

String 전체에서 해당하는 것 다 찾아줌

규칙을 어떻게 넣느냐에 따라 엄청나게 smart하게 찾을 수 있음

In [5]:
re.sub('a', 'b', 'abcdefa')

'bbcdefb'

substitute한다는 의미임 (sub)

대상이 되는 string 'abcdefa'에서 'a'를 'b'로 바꾼다 (전부 다에 해당됨) a를 전부 다 b로 바꿔주는 function

In [ ]:
'''       Basic Regular Expression Meta-Characters, Including Wildcards, Ranges and Closures

.	        Wildcard, matches any character
^abc	    Matches some pattern abc at the start of a string
abc$	    Matches some pattern abc at the end of a string
[abc]	    Matches one of a set of characters
[^abc]    Matches anything but a set of characters
[A-Z0-9]	Matches one of a range of characters
ed|ing|s	Matches one of the specified strings (disjunction)
*	        Zero or more of previous item, e.g. a*, [a-z]* (also known as Kleene Closure)
+	        One or more of previous item, e.g. a+, [a-z]+
?	        Zero or one of the previous item (i.e. optional), e.g. a?, [a-z]?
{n}	      Exactly n repeats where n is a non-negative integer
{n,}	    At least n repeats
{,n}	    No more than n repeats
{m,n}	    At least m and no more than n repeats
a(b|c)+	  Parentheses that indicate the scope of the operators
(...)     Matches whatever regular expression is inside the parentheses
\d
Matches any decimal digit; this is equivalent to the class [0-9].
\D
Matches any non-digit character; this is equivalent to the class [^0-9].
\s
Matches any whitespace character; this is equivalent to the class [ \t\n\r\f\v].
\S
Matches any non-whitespace character; this is equivalent to the class [^ \t\n\r\f\v].
\w
Matches any alphanumeric character; this is equivalent to the class [a-zA-Z0-9_].
\W
Matches any non-alphanumeric character; this is equivalent to the class [^a-zA-Z0-9_].

'''

In [ ]:
nltk.download('words')
engdict = nltk.corpus.words.words('en')
engdict

List로 23만 개의 단어가 들어옴 len(engdict)하면 나오는 것

각각의 string에 대하여 search와 findall은 어디에 적용되냐면, 'string'에 적용되는 것임 -> 그래서 밑에서 list로 engdict가 받아짐

각각의 아이템 (ex. 0번째 item)

In [9]:
engdict[0]

'A'

각각의 아이템에 대해 search와 findall을 쓸 것임!

In [11]:
engdict = nltk.corpus.words.words('en')

result = [w for w in engdict if re.search('ed$', w)]
# result = [w for w in engdict if re.search('^..j..t..$', w)]
# result = [w for w in engdict if re.search('^[ghi][mno][jlk][def]$', w)]
# result = [w for w in engdict if re.search('^[ah]+$', w)][:10]
print(result[:10])

['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']


engdict라는 list가 있는데, list comprehension 잘 이용해서 (for w in engdict) engdict에 있는 각각의 아이템을 하나씩 w로 받아서 그 개수만큼 forloop를 돈다!

각 loop에 들어오는 (w로 들어오는) 그게 if절을 통과하게 됨. 

w로 들어와서 'regular expression search'에 들어오게 됨

해당되는 string에 대하여 'ed$'를 찾아라 = 어떤 식으로의 규칙임 (regular expression) 나름대로의 규칙 적용시킨 것임

첫 번째 줄에서 if에서 re.search가 w의 저 string 부분이 있으면 (뭐라도 있으면) w에 담기는 것이고, a가 w에 없다면 if가 그냥 지나치는 것임

if는 뭐라도 잡아오면 통과되어서 w에 listing되는 것임

**'ed$': string의 끝에서 이런 패턴이 있는가?** -> String의 끝에서 'ed'라는 패턴이 있는가?

string의 시작에서는 ^라는 기호를 씀. -> 시작부에서 이런 패턴이 있는가?

re.search라는 것은 이런 string에서 이런 규칙이 발견되면 뱉어주는 것을 의미함

In [10]:
re.search('ed$', 'educated')

<re.Match object; span=(6, 8), match='ed'>

string 끝에 있는 'ed'를 search해서 가져와라!

그냥 ed라고 하면 제일 처음 hit하는 'ed'를 가져옴.

^ed하면 string의 제일 앞에서 나오는 'ed'를 명시적으로 가져오게 됨 

$ 표시 어떻게 쓰고 (string의 끝이라는 뜻)

^ 표시 어떻게 쓰는지 (string의 시작)

잘 파악하기

string 제일 뒤에 'ed'로 끝나는 게 있으면 search가 'ed'를 가져오게 됨. 그러면 통과! (ed를 가져왔기 때문에 통과)

가져오지 못했으면 그냥 지나치는 것임

결과적으로 list에 쌓이는 것은 'engdict'에서 'ed'로 끝나는 단어들만 쌓이는 것임

In [12]:
len(result)

9192

len(engdict)가 23만 개 였는데

result에서 'ed'로 끝나는 걸 싹 다 받았더니 9192개가 들어온 것임

한 줄로 사전 전체를 search해서 영어 단어에서 'ed'로 끝나는 게 몇 개 있는지 클릭 하나로 해결된 것임. 왜 우리가 코딩 배워야 하는지 이유가 여기 있죠? 사람이 스스로 했다고 생각하면 양이 너무 많음...



In [13]:
engdict = nltk.corpus.words.words('en')

# result = [w for w in engdict if re.search('ed$', w)]
result = [w for w in engdict if re.search('^..j..t..$', w)]
# result = [w for w in engdict if re.search('^[ghi][mno][jlk][def]$', w)]
# result = [w for w in engdict if re.search('^[ah]+$', w)][:10]
print(result[:10])

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector']


중간중간에 점을 찍은 건 뭘 의미할까?

wildcard, matches any character: . 하나면 한 캐릭터

중간중간 글자들로 채워지는 것임. 그런 글자들만 한 번 찾아봐라! (총 9개 나옴)

23만 개 중에서 이 조건을 만족하는 애들을 순식간에 찾아줌. 이 조건이 너무 powerful해서 우리가 생각하는 거 다 할 수 있음

In [15]:
len(result)

13

In [16]:
engdict = nltk.corpus.words.words('en')

# result = [w for w in engdict if re.search('ed$', w)]
# result = [w for w in engdict if re.search('^..j..t..$', w)]
result = [w for w in engdict if re.search('^[ghi][mno][jlk][def]$', w)]
# result = [w for w in engdict if re.search('^[ah]+$', w)][:10]
print(result[:10])

['gold', 'golf', 'hold', 'hole']


[]대괄호 뜻이 무엇일까? Matches one of a set of characters: 대괄호 속에 들어 있는 abc 중 하나를 만족시키는 것 (ghi 중에 하나) 그 다음 캐릭터는 mno 중에 하나, 그 다음 캐릭터 jlk, 그 다음은 def 캐릭터 중 하나

총 4개의 캐릭터로 되어 있다는 것을 이해해야 함

In [18]:
engdict = nltk.corpus.words.words('en')

# result = [w for w in engdict if re.search('ed$', w)]
# result = [w for w in engdict if re.search('^..j..t..$', w)]
# result = [w for w in engdict if re.search('^[ghi][mno][jlk][def]$', w)]
result = [w for w in engdict if re.search('^[ah]+$', w)]
print(result[:10])

['a', 'aa', 'ah', 'aha', 'h', 'ha', 'hah']


a 혹은 h의 캐릭터가 +$ 

+가 의미하는 것: one or more of previous item: a+되면 a가 한 번 이상 나와야 한다는 의미임

a 혹은 h가 한 번 이상 나오는 걸 다 찾아라! (조건 만족하는 거 7개 밖에 없음)

In [20]:
nltk.download('treebank')
wsj = nltk.corpus.treebank.words()
wsj

[nltk_data] Downloading package treebank to /root/nltk_data...
[nltk_data]   Package treebank is already up-to-date!


['Pierre', 'Vinken', ',', '61', 'years', 'old', ',', ...]

똑같은 것을 'treebank'에서 해 보기.

wsj라고 해서 신문에 있는 단어 (wallstreet journal): 비정형적인 단어도 있을 수 있음 (ex. $1200) 사전이랑 조금 다른, 현실과 훨씬 가까운 단어 corpus라고 생각하면 됨

In [21]:
result = [w for w in wsj if re.search('(ed|ing)$', w)]
# result = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
# result = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
# result = [w for w in wsj if re.search('^[0-9]{4}$', w)]
# result = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
# result = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]

result = sorted(set(result))
print(result[:10])

['62%-owned', 'Absorbed', 'According', 'Adopting', 'Advanced', 'Advancing', 'Alfred', 'Allied', 'Annualized', 'Anything']


wsj list에서 search를 통화하면 'ed'가 있기만 하면 통과해서 쌓인다! 라고 이해하면 될 것 같음

(ed|ing)$: 괄호가 되어 있고, 그 사이에 bar이 있음. 얜 무슨 의미?
-> ed 혹은 ing 혹은 s 이런 뜻임. (Bar의 의미)

괄호는 그냥 묶어준 것이고, ed 혹은 ing로 끝나는 단어를 모아서 재정의해라!

sorted가 들어가면 sort해주는 작업 나오고, 저 줄이 없으면 그냥 신문에 나오는 순서대로 막 나옴

In [22]:
# result = [w for w in wsj if re.search('(ed|ing)$', w)]
result = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
# result = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
# result = [w for w in wsj if re.search('^[0-9]{4}$', w)]
# result = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
# result = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]

result = sorted(set(result))
print(result[:10])

['0.0085', '0.05', '0.1', '0.16', '0.2', '0.25', '0.28', '0.3', '0.4', '0.5']


[0-9]: '숫자'라는 뜻임 (숫자를 저렇게 씀)

숫자가 하나 이상 나온다 [0-9]+

역 slash.: 이게 '.'임. ...은 그냥 'any character'을 의미함. 실제 .찍고 싶으면 /. 이렇게 찍음 

또 숫자가 하나 이상 나오고 끝나는 것을 찾아라
-> 소수점이 들어가 있는 숫자를 의미함. 그래서 그런 숫자들이 쭉 나옴

In [23]:
# result = [w for w in wsj if re.search('(ed|ing)$', w)]
# result = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
result = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
# result = [w for w in wsj if re.search('^[0-9]{4}$', w)]
# result = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
# result = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]

result = sorted(set(result))
print(result[:10])

['C$', 'US$']


/$가 실제 달러 표시를 의미함. 그냥 단독으로 $ 쓰면 이미 string 끝이라고 쓰였기 때문에, 실제 달러 표시 쓰고 싶으면 /$로 구분해 주는 것임 (역슬래쉬+$)

A-Z인데, 어떠한 '대문자'를 의미함. 어떠한 대문자가 하나 이상 나오고, $로 끝나면 다 찾아와라.

10만 개 중에서 2개밖에 없음

In [ ]:
# result = [w for w in wsj if re.search('(ed|ing)$', w)]
# result = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
# result = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
result = [w for w in wsj if re.search('^[0-9]{4}$', w)]
# result = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
# result = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]

result = sorted(set(result))
print(result[:10])

String의 시작, 0에서 9까지의 숫자가 {4} 몇 번 나오는가를 의미함 (준괄호 앞에 있는 게 몇 번 나오는가?

**5:36~ 다시 듣기**



In [25]:
# result = [w for w in wsj if re.search('(ed|ing)$', w)]
# result = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
# result = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
# result = [w for w in wsj if re.search('^[0-9]{4}$', w)]
result = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
# result = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]

result = sorted(set(result))
print(result[:10])

['10-day', '10-lap', '10-year', '100-share', '12-point', '12-year', '14-hour', '15-day', '150-point', '190-point']


In [ ]:
import os
url = "https://raw.githubusercontent.com/hsnam95/my/main/friends_season01_script.txt"
os.system("curl " + url + " > friends_season01_script.txt")

# read a text file in the server
file = open("friends_season01_script.txt")
text = file.read()
file.close()
text

In [28]:
pattern = '(?<=: ).+(?=[\.|\?|\!])'
sent = re.findall(pattern, text)
sent
text = '\n'.join(sent)
text

"how much Barry looks like Mr. Potato Head. Y'know, I mean, I always knew looked familiar, but... Anyway, I just had to get out of there, and I started wondering 'Why am I doing this, and who am I doing this for?'\nWait\nBring back the comedian\nthat is not such a terrible thing\nRight hand blue. (Phoebe has to bend over\nall the fun, all the talking, all the sex; and none of the responsibility\nyou dumped her. Right? I mean, this woman was unbelievably sexy, and beautiful, intelligent, unattainable..\nIf Jill Goodacre offers you gum, you take it\ndesperate scramble to find anything with lips just so you can have someone to kiss when the ball drops!\nthe always interesting Nora Tyler Bing\nRachel Green\nif youÕre ever in a similar situation, never ever leave your hand... (he mimes Nina taking her revenge) .\nYeah, um..\nReally? (out loud): Sure, we have scones left\nDear Ms. Green, thank you for your inquiry, however... oh.\nWe have apple cinnamon.\nOK... Dear Ms. Green... yeah... yeah

나의 왼쪽, 오른쪽이 어떤 환경인가? 

하나 이상의 캐릭터가 있고 오른쪽 환경은 세 가지 punctuation 중 하나를 찾아라!

아주 긴 것 중에서 이런 것 만족시키는 것 list화 해라 (sent)

분리했으니까 영어교재로 쓰든 뭐 어떻게 하든 사용하기 훨씬 쉬운 것임

줄바꿈을 가지고 다 붙여라 (긴 string이 다시 만들어지는 것임) 이 작업을 할 수 있어야 함

줄바꿈으로 해서 전부 다 결합시켜야 함

' '이렇게 하면 한 칸 띄움으로 다 연결되는 것 의미함

In [29]:
# write out a text file
file = open("tmp.txt", "w")
file.write(text)
file.close()

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

given string이 있을 때, 그걸 write out할 수 있음

그것을 저렇게 쓰면 나감 (download할 수 있다는 장점이 있음)



오늘 숙제: 주어진 text가 이렇게 있음. 오늘 배운 text가 friends season 1 text임

남학생은 'Ross' 부분, 여학생은 여기 나오는 'Rachel'이라고 나오는 이름을 모두 여러분들의 학번으로 바꾸세요. (번호가 나오겠죠?)

바꿔서 Text로 저장하기 (20220509) 

하나는 20220509.ipynb 파일 하나 꼭 만들고,

그 파일 속에는 방금 제가 얘기한 게 코드로 들어가 있으면 됩니다. 코드 결과로 나오는 게 20220509.txt가 됩니다.

re.sub 써서 해라!!

그걸 썼다는 것을 code ipymb로 만들어야 함

"friends" 영화 파일에서 Rachel을 우리 학번으로 바꿔서 아웃풋 만들어 내는 거 해봐라.